create_library_to_link: add mangle_name and unique_name arguments

aherrmann opened the pull request:
Closes #8180 

- Adds `mangle_name` and `unique_name` arguments to `create_library_to_link`
    - `mangle_name` allows users replicate the behavior of `cc_library` in their custom rules. The link placed under the solib directory will be mangled and no intermediate directory will be generated.
    - `unique_name` allows users to specify that the given library name is unique and no intermediate directory is required.
- Adds the `prefixLibraryPath` argument to the internal `CcCommon` API to facilitate the above changes to `create_library_to_link`.
- Adds shell tests for the `mangle_name` and `unique_name` arguments to `create_library_to_link`.

### Motivation

See https://github.com/bazelbuild/bazel/issues/8180 https://github.com/tweag/rules_haskell/issues/958

Before this PR the new Starlark Cc API's `create_library_to_link` would always preserve the given library name and link it into the solib directory underneath an intermediate directory based on the library label and consumer label. These intermediate directories can grow very long, in particular if external workspaces are involved. In rules_haskell this causes issues on Windows and MacOS. On Windows the linker cannot look for libraries underneath search paths which exceed a certain path length. On MacOS the header size of a MACH-O binary/library is limited and the required runpaths/load commands would quickly exceed this limit due to the long paths generated by Bazel. In rules_haskell in particular these intermediate directories are unnecessary, as rules_haskell already mangles the library names itself to ensure their uniqueness. Furthermore, the Haskell compiler GHC expects library names of a particular shape, such that Bazel's builtin mangling cannot be used. This motivates the combination `mangle_name = False`, `unique_name = True` enabled by this PR.

cc @oquenchil @mboes @cocreature

aherrmann commented:
CI has passed and this is now ready for review.

The tests test the symlinks generated in the solib directory. Since these are not generated on Windows, I have disabled the tests on Windows.

mboes commented:
This PR just exposes existing method parameters that are there in the Java API to Starlark code. It's currently a blocker for a few issues in rules_haskell.

oquenchil commented:
I will take a look at this on Monday next week. Even though this is in Java, it doesn't mean we want to expose it to Starlark. We are carefully reviewing every parameter we add to the API because unlike in Java fixing a mistake there is not as simple as an internal refactoring of our code.

oquenchil commented:
So if I understand correctly, you want more flexibility in the name used for the symlink created. I think better than the two booleans on the API, the change should add a String parameter which internally gets converted to a `PathFragment`. Then internally override the method `getDynamicLibrarySymlink` with a variation that doesn't take `preserveName` and `prefixConsumer`, but takes that `PathFragment` instead and where `getMangledNamed()` is not called.

That should give enough flexibility to do what you want. But first I have to ask, why was it causing issues to have the default `preserveName=true` and `prefixConsumer=true`. For a given instance, what was the name of the symlink produced and what was the name that you wanted and **why** did you want that different name?

Regarding the test, I would prefer if you remove the shell test altogether and add a new test here:
https://github.com/bazelbuild/bazel/blob/master/src/test/java/com/google/devtools/build/lib/rules/cpp/SkylarkCcCommonTest.java

Thanks for this PR!

aherrmann commented:
> So if I understand correctly, you want more flexibility in the name used for the symlink created.

Yes, that is correct.

> I think better than the two booleans on the API, the change should add a String parameter which internally gets converted to a `PathFragment`. Then internally override the method `getDynamicLibrarySymlink` with a variation that doesn't take `preserveName` and `prefixConsumer`, but takes that `PathFragment` instead and where `getMangledNamed()` is not called.

Interesting, that sounds feasible, I'll look into it.

> But first I have to ask, why was it causing issues to have the default `preserveName=true` and `prefixConsumer=true`. For a given instance, what was the name of the symlink produced and what was the name that you wanted and **why** did you want that different name?

Sure, take tweag/rules_haskell@21c277b5a7ef019bbd10348568538723aef2699c with Bazel 0.27.0 on Linux with the default GHC bindist. Then run `bazel build //tests/binary-with-lib-dynamic`.
This will produce
```
$ ls bazel-bin/_solib_k8/*
bazel-bin/_solib_k8/_U@rules_Uhaskell_S_Shaskell_Ctoolchain-libraries___Uexternal_Srules_Uhaskell_Ughc_Ulinux_Uamd64_Slib_Sbase-4.12.0.0:
libHSbase-4.12.0.0-ghc8.6.5.so

bazel-bin/_solib_k8/_U@rules_Uhaskell_S_Shaskell_Ctoolchain-libraries___Uexternal_Srules_Uhaskell_Ughc_Ulinux_Uamd64_Slib_Sghc-prim-0.5.3:
libHSghc-prim-0.5.3-ghc8.6.5.so

bazel-bin/_solib_k8/_U@rules_Uhaskell_S_Shaskell_Ctoolchain-libraries___Uexternal_Srules_Uhaskell_Ughc_Ulinux_Uamd64_Slib_Sinteger-gmp-1.0.2.0:
libHSinteger-gmp-1.0.2.0-ghc8.6.5.so

bazel-bin/_solib_k8/_U@rules_Uhaskell_S_Shaskell_Ctoolchain-libraries___Uexternal_Srules_Uhaskell_Ughc_Ulinux_Uamd64_Slib_Srts:
libffi.so  libHSrts-ghc8.6.5.so

bazel-bin/_solib_k8/_U_S_Stests_Sbinary-with-lib-dynamic_Clib___Utests_Sbinary-with-lib-dynamic:
libHStestsZSbinary-with-lib-dynamicZSlib-ghc8.6.5.so
```
By default Bazel produces very long directory names, and separate directories for most libraries. On macOS, the standard GHC bindist requires dynamic linking as part of compilation, each of these long paths will have to appear in the MACH-O header as a runpath or load command. This can very quickly add up to exceed the MACH-O header size limit. The header may not exceed 32k. With paths like the above we have frequently had issues with builds failing due to MACH-O header size at around 200 library dependencies. Currrently, rules_haskell employs a number of workarounds to avoid the issue. Shortening the symlink paths as with this PR would make our lives as rules_haskell maintainers easier, and also raise the upper bound for the number of possible library dependencies.

For core Haskell libraries as well as `haskell_library` targets, the directory prefix is unnecessary. The first few libraries are core libraries shipped with the compiler toolchain. Their paths look like this:
```
_U@rules_Uhaskell_S_Shaskell_Ctoolchain-libraries___Uexternal_Srules_Uhaskell_Ughc_Ulinux_Uamd64_Slib_Sbase-4.12.0.0/libHSbase-4.12.0.0-ghc8.6.5.so
```
and should instead look like this
```
libHSbase-4.12.0.0-ghc8.6.5.so
```
The library name contains the name and version of the corresponding core package and the compiler version. This makes accidental name collisions very, very unlikely.

The last library comes from a local `haskell_library` target. Its path looks like this:
```
_U_S_Stests_Sbinary-with-lib-dynamic_Clib___Utests_Sbinary-with-lib-dynamic/libHStestsZSbinary-with-lib-dynamicZSlib-ghc8.6.5.so
```
and should instead look like this
```
libHStestsZSbinary-with-lib-dynamicZSlib-ghc8.6.5.so
```
The library name already contains the full package path and is thereby unique. The `libHS` prefix is required by the Haskell compiler, same as the compiler version suffix.

> Regarding the test, I would prefer if you remove the shell test altogether and add a new test here:
> https://github.com/bazelbuild/bazel/blob/master/src/test/java/com/google/devtools/build/lib/rules/cpp/SkylarkCcCommonTest.java

Right, I'm not too familiar with the internal Java API. If I understand correctly, I'd generate the necessary files similar to how it's done in [`setUpCcLinkingContextTest`](https://github.com/bazelbuild/bazel/blob/12ebb84a7645baffdbaf9f40cbfcec996a540ffe/src/test/java/com/google/devtools/build/lib/rules/cpp/SkylarkCcCommonTest.java#L1346) and then test the symlinks similar to how it's done [here](https://github.com/bazelbuild/bazel/blob/12ebb84a7645baffdbaf9f40cbfcec996a540ffe/src/test/java/com/google/devtools/build/lib/rules/cpp/SkylarkCcCommonTest.java#L1258-L1275). Correct?

Thanks for the review @oquenchil

aherrmann commented:
@oquenchil I've addressed your comments. `create_library_to_link` now takes optional arguments `dynamic_library_path` and `interface_library_path`. If set, these override the path of the symbolic link underneath the solib directory. I've also removed the shell integration tests and instead added Java tests in `SkylarkCcCommonTest.java` as requested. PTAL.

oquenchil commented:
Yes, that looks perfect. One little nit though. The `dynamic_library_path` and `interface_library_path` are changing the paths only for the symlinks correct? Ideally, these two names would make that clear, so something like `dynamic_library_symlink_path` and `interface_library_symlink_path`.

Otherwise, I think it might cause confusion with people thinking that the path of the actual `dynamic_library` and `interface_library` gets replaced.

Also initially I wanted to complain about symlinks being an internal optimization detail which is platform dependent and shouldn't be leaked. But this is already leaking through the `resolved_symlink_dynamic_library` and `resolved_symlink_dynamic_library` getters, so no issue here.

The PR looks perfect, the test is exactly how it should be and I appreciate you took good care of limiting the filetypes for these two parameters. Just renaming the two input parameters to make clear that it's for the symlink and we can merge this.

aherrmann commented:
@oquenchil Thank you, that's great to hear!

> The dynamic_library_path and interface_library_path are changing the paths only for the symlinks correct? Ideally, these two names would make that clear, so something like dynamic_library_symlink_path and interface_library_symlink_path.

Yes, that is correct. Agreed, makes sense to be more explicit in those names. I've changed the names as suggested. PTAL

aherrmann commented on src/main/java/com/google/devtools/build/lib/skylarkbuildapi/cpp/CcModuleApi.java:
```diff
@@ -586,6 +586,24 @@ CcToolchainVariablesT getLinkBuildVariables(
+        @Param(
+            name = "dynamic_library_symlink_path",
+            doc =
+                "Override the default path of the dynamic library link in the solib directory.",
```
I've amended the docstring following https://github.com/bazelbuild/bazel/pull/8888#discussion_r318149313. That should clarify this a bit more. Also, the html documentation will display the type `String` above that description, similar to [here](https://docs.bazel.build/versions/0.28.0/skylark/lib/cc_common.html#compile.name).

aherrmann commented on src/main/java/com/google/devtools/build/lib/skylarkbuildapi/cpp/CcModuleApi.java:
```diff
@@ -586,6 +586,24 @@ CcToolchainVariablesT getLinkBuildVariables(
+            named = true,
+            noneable = true,
+            type = String.class,
+            defaultValue = "None"),
```
You're right that an empty string is not meaningful here. I was trying to follow the convention of `None` meaning "use default behavior". But, it looks like for strings `""` is indeed used for that purpose, e.g. in `repository_ctx.execute(..., working_directory = "")`, so this change seems alright. The advantage is that it makes it easier to forward a default value from a rule string attribute, like [here](https://github.com/DACH-NY/bazel/commit/e89bddce772dae42c1bf59437bcf4c903995ba59#diff-8973c7070089ad4560f3d03301033b12L1493-R1494).

aherrmann commented:
@c-parsons @oquenchil Can this be merged?

c-parsons commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CcModule.java:
```diff
@@ -386,9 +390,38 @@ public LibraryToLink createLibraryLinkerInput(
+    PathFragment dynamicLibraryPathFragment = null;
+    if (dynamicLibraryPath != null && !dynamicLibraryPath.isEmpty()) {
+      dynamicLibraryPathFragment = PathFragment.create(dynamicLibraryPath);
+      if (dynamicLibraryPathFragment.isEmpty() || dynamicLibraryPathFragment.isAbsolute() || dynamicLibraryPathFragment.containsUplevelReferences()) {
```
Do we need to check dynamicLibraryPathFragment.isEmpty(), given that we already checked dynamicLibraryPath.isEmpty above?

aherrmann replied:
Good question. `PathFragment.create` performs normalization if necessary. The docs don't seem to guarantee that normalization never returns an empty `PathFragment` given a non-empty `String`. E.g. `./` or `foo/..` could potentially be normalized away. For reference, the docs say

```
 * <p>Strings are normalized with '.' and '..' removed and resolved (if possible), any multiple
 * slashes ('/') removed, and any trailing slash also removed. Windows drive letters are uppercased.
 * The current implementation does not touch the incoming path string unless the string actually
 * needs to be normalized.
```

c-parsons commented on src/test/java/com/google/devtools/build/lib/rules/cpp/SkylarkCcCommonTest.java:
```diff
@@ -1275,6 +1275,76 @@ public void testIncompatibleDepsetForLibrariesToLinkGetter() throws Exception {
+    assertThat(
+            librariesToLink.toCollection(LibraryToLink.class).stream()
+                .filter(x -> x.getDynamicLibrary() != null)
+                .map(x -> x.getDynamicLibrary().getRootRelativePath().relativeTo(toolchain.getSolibDirectory()).toString())
```
nit: can you refactor x.getDynamicLibrary().getRootRelativePath().relativeTo(toolchain.getSolibDirectory()).toString() to a helper private static method, to make this slightly more readable? (the combination of chained calls and use of stream() makes this difficult to read)

aherrmann replied:
I've factored it out as `getSolibRelativePath`.

oquenchil commented:
> Sorry this review fell off my radar In general, this looks good to me. But I'd like to get approval from oquenchil@ before proceeding, as he has more knowledge of this piece of the API.

Yes, still looks good to me.

mjrussell commented:
Im running into MACH-O limits in rules_haskell and would greatly appreciate this getting merged. Is there anything holding this up from merging?

aherrmann commented:
@c-parsons I think this is ready for merge. Let me know if there's something else you'd like to see changed in this PR.

@mjrussell Thanks for pinging. Regarding MACH-O limits, which version of rules_haskell are you on? [Version 0.11](https://github.com/tweag/rules_haskell/releases/tag/v0.11) contains some workarounds that should avoid most MACH-O header size issues. If you're still hitting these issues on 0.11, could you open an issue at https://github.com/tweag/rules_haskell with some more details?

c-parsons commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CcModule.java:
```diff
@@ -386,9 +390,38 @@ public LibraryToLink createLibraryLinkerInput(
+      if (!Link.ONLY_SHARED_LIBRARY_FILETYPES.matches(dynamicLibraryPathFragment.getBaseName())) {
+        String extensions = Link.ONLY_SHARED_LIBRARY_FILETYPES.toString();
+        extensionErrorsBuilder.append(
+            String.format("'%s' %s %s", dynamicLibraryPathFragment.toString(), extensionErrorMessage, extensions));
```
There's no need to call toString() on objects within String.format(), as it already does the appropriate conversion
(There are also 3 other instances of this in this file)

aherrmann replied:
Good point, I've removed the calls to `.toString()`.

c-parsons commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CcModule.java:
```diff
@@ -386,9 +390,38 @@ public LibraryToLink createLibraryLinkerInput(
+    }
+
+    PathFragment interfaceLibraryPathFragment = null;
+    if (interfaceLibraryPath != null && !interfaceLibraryPath.isEmpty()) {
```
This looks a lot like the validation above. Refactor into a helper method?

aherrmann replied:
Indeed, they're very similar. I've factored them out as `validateSymlinkPath`.

c-parsons commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CcModule.java:
```diff
@@ -386,9 +390,38 @@ public LibraryToLink createLibraryLinkerInput(
     String extensionErrorMessage = "does not have any of the allowed extensions";
+
+    PathFragment dynamicLibraryPathFragment = null;
+    if (dynamicLibraryPath != null && !dynamicLibraryPath.isEmpty()) {
```
Use Strings.isNullOrEmpty() instead  (from Guava)
https://guava.dev/releases/23.0/api/docs/com/google/common/base/Strings.html#isNullOrEmpty-java.lang.String-

aherrmann replied:
I've replaced these checks by calls to `Strings.isNullOrEmpty()`.

c-parsons commented on src/main/java/com/google/devtools/build/lib/rules/cpp/SolibSymlinkAction.java:
```diff
@@ -131,6 +131,37 @@ public static Artifact getDynamicLibrarySymlink(
+      PathFragment path) {
+    Preconditions.checkArgument(Link.SHARED_LIBRARY_FILETYPES.matches(library.getFilename()));
+    Preconditions.checkArgument(Link.SHARED_LIBRARY_FILETYPES.matches(path.getBaseName()));
+    Preconditions.checkArgument(!library.getRootRelativePath().getSegment(0).startsWith("_solib_"));
```
Is this validated anywhere in the caller? (I did not see this in my visual inspection)
If any of these Precondition checks fail, it will cause a Bazel crash. We should avoid cases where bad user input can result in a Bazel crash -- we should instead throw the (checked) EvalException in such cases.

aherrmann replied:
The filetypes are already checked in the calling code. I've added a check for the `_solib_` path precondition.

aherrmann commented:
@c-parsons I've addressed your comments. PTAL.

c-parsons commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CcModule.java:
```diff
@@ -396,9 +400,22 @@ public LibraryToLink createLibraryLinkerInput(
 
-    Artifact notNullArtifactForIdentifier = null;
     StringBuilder extensionErrorsBuilder = new StringBuilder();
     String extensionErrorMessage = "does not have any of the allowed extensions";
```
dead code

aherrmann replied:
It's still used in the validation logic for the static and dynamic library attributes. E.g. [here](https://github.com/bazelbuild/bazel/pull/8888/files/43dc9a749595ab764c65eb7dd42122e2b5460899#diff-f321a84e135b6943c822c7a94df47b82R428). Sadly, it doesn't match so well with the symlink validation and refactoring that code too much seems out of scope for this PR.

c-parsons commented on src/main/java/com/google/devtools/build/lib/rules/cpp/CcModule.java:
```diff
@@ -507,6 +562,24 @@ public LibraryToLink createLibraryLinkerInput(
   }
 
+  private static void validateSymlinkPath(Location location, String attrName, PathFragment symlinkPath, FileTypeSet filetypes, StringBuilder errorsBuilder) throws EvalException {
+    String extensionErrorMessage = "does not have any of the allowed extensions";
```
please just inline this string

aherrmann replied:
done

aherrmann commented:
> Will merge shortly

Awesome, thank you!
